src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs

```diff
@@ -220,6 +227,89 @@ private ExtensionReference[] ParseExtensions(string metadataFilePath)
+                var minimumVersion = new Version(requirement.MinimumVersion);
+
+                if (bundleVersion < minimumVersion)
+                {
+                    _logger.MinimumBundleVersionNotSatisfied(bundleDetails.Id, bundleDetails.Version, requirement.MinimumVersion);
+                    throw new HostInitializationException($"Referenced bundle {bundleDetails.Id} of version {bundleDetails.Version} does not meet the required minimum version of {requirement.MinimumVersion}. Update your extension bundle reference in host.json to reference {requirement.MinimumVersion} or later. For more information see https://aka.ms/func-min-bundle-versions.");
+                }
+            }
+        }
+
+        private void ValidateExtensionRequirements(List<Type> startupTypes)
+        {
+            var errors = new List<string>();
+
+            void CollectError(Type extensionType, Version minimumVersion, ExtensionStartupTypeRequirement requirement)
+            {
+                _logger.MinimumExtensionVersionNotSatisfied(extensionType.Name, extensionType.Assembly.FullName, minimumVersion, requirement.PackageName, requirement.MinimumPackageVersion);
+                string requirementNotMetError = $"ExtensionStartupType {extensionType.Name} from assembly '{extensionType.Assembly.FullName}' does not meet the required minimum version of {minimumVersion}. Update your NuGet package reference for {requirement.PackageName} to {requirement.MinimumPackageVersion} or later.";
+                errors.Add(requirementNotMetError);
+            }
+
+            foreach (var extensionType in startupTypes)
+            {
+                if (_extensionRequirements.ExtensionRequirementsByStartupType.ContainsKey(extensionType.Name))
+                {
+                    ExtensionStartupTypeRequirement requirement = _extensionRequirements.ExtensionRequirementsByStartupType[extensionType.Name];
+                    Version minimumAssemblyVersion = new Version(requirement.MinimumAssemblyVersion);
+
+                    Version extensionAssemblyVersion = extensionType.Assembly.GetName().Version;
+                    string extensionAssemblySimpleName = extensionType.Assembly.GetName().Name;
+
+                    if (extensionAssemblySimpleName != requirement.AssemblyName)
+                    {
+                        // We recognized this type, but it did not come from the assembly that we expect, skipping validation
+                        continue;
+                    }
+
+                    if (requirement.MinimumAssemblyFileVersion != null)
+                    {
+                        Version minimumAssemblyFileVersion = new Version(requirement.MinimumAssemblyFileVersion);
+
+                        try
+                        {
+                            Version extensionAssemblyFileVersion = new Version(System.Diagnostics.FileVersionInfo.GetVersionInfo(extensionType.Assembly.Location).FileVersion);
```

Comment:
This might be an expensive call in the startup path. Do we want to support file versions? I'll look at the file, but I'm assuming this would be for a small number of scenarios, correct?

Reply:
No concrete scenarios today, but in dynamic assembly scenarios (and some load scenarios like loading from bytes), `Location` will not be populated as expected, so it would be good to guard against that.

Reply:
We need to support it to provide the right check for durable. Notice that I coded this path to only activate if a file version specifier is present - durable is the only one with that specifier, and so we will only run through this codepath if the durable extension is present.

Reply:
I did add a guard for the location though

Reply:
I did notice the check for the file version and did see that Durable is the only one with that set, but I do worry about the impact as it will be present in a large number of apps.